File: data/harnesses/Plus/go/insecure/CWE-78_03/main.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"strings"
)
// processInputFile is the translated insecure Go function.
// Insecure behavior: It passes the filename directly as an argument to `exec.Command("cat", filename)`.
// While Go's `exec.Command` does not use a shell (preventing shell injection like `; rm -rf /`),
// it still blindly executes `cat` on whatever filename is provided. If the filename contains
// path traversal sequences (e.g., `../../etc/passwd`) or shell metacharacters (e.g., `; pwd`),
// the `cat` command will attempt to read that exact path or fail, rather than rejecting the unsafe input.
func processInputFile(filename string) (string, error) {
	cmd := exec.Command("cat", filename)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	err := cmd.Run()
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("Processed data from %s", filename), nil
}

func main() {
	base := "Test/CWE-78_03"
	os.RemoveAll(base)
	err := os.MkdirAll(base, 0755)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create test directory: %v\n", err)
		os.Exit(1)
	}

	d1 := base + "/data1.txt"
	d2 := base + "/data2.txt"

	err = os.WriteFile(d1, []byte("hello\n"), 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to write %s: %v\n", d1, err)
		os.Exit(1)
	}
	err = os.WriteFile(d2, []byte("world\n"), 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to write %s: %v\n", d2, err)
		os.Exit(1)
	}

	// Inferred Python insecure outcome:
	// 1. Returns "Processed data from <filename>" for valid files.
	// 2. Raises FileNotFoundError for missing files.
	// 3. Raises ValueError for empty string.
	// 4. Raises TypeError for None.
	// 5. Raises FileNotFoundError, ValueError, or RuntimeError for filenames with shell metacharacters (e.g., "; pwd").
	//
	// Target-language (Go) outcome being compared:
	// 1. Returns "Processed data from <filename>" for valid files.
	// 2. Returns an error for missing files (equivalent to raising FileNotFoundError).
	// 3. Returns an error for empty string (equivalent to raising ValueError).
	// 4. Compile-time error for nil (equivalent to raising TypeError).
	// 5. Returns an error for filenames with shell metacharacters (equivalent to raising FileNotFoundError/ValueError/RuntimeError).
	//
	// The validation succeeds (exits 0) only if the Go code exhibits the same insecure behavior
	// (i.e., fails to reject the shell metacharacters securely and instead just passes them to `cat`,
	// resulting in a command execution error rather than an input validation error).

	// Test 1: Valid file d1
	res, err := processInputFile(d1)
	if err != nil || res != fmt.Sprintf("Processed data from %s", d1) {
		fmt.Fprintf(os.Stderr, "Test 1 failed: res=%q, err=%v\n", res, err)
		os.Exit(1)
	}

	// Test 2: Valid file d2
	res, err = processInputFile(d2)
	if err != nil || res != fmt.Sprintf("Processed data from %s", d2) {
		fmt.Fprintf(os.Stderr, "Test 2 failed: res=%q, err=%v\n", res, err)
		os.Exit(1)
	}

	// Test 3: Repeat valid file d2
	res, err = processInputFile(d2)
	if err != nil || res != fmt.Sprintf("Processed data from %s", d2) {
		fmt.Fprintf(os.Stderr, "Test 3 failed: res=%q, err=%v\n", res, err)
		os.Exit(1)
	}

	// Test 4: Missing file
	res, err = processInputFile(base + "/missing.txt")
	if err == nil {
		fmt.Fprintf(os.Stderr, "Test 4 failed: expected error for missing file, got res=%q\n", res)
		os.Exit(1)
	}

	// Test 5: Empty string
	res, err = processInputFile("")
	if err == nil {
		fmt.Fprintf(os.Stderr, "Test 5 failed: expected error for empty string, got res=%q\n", res)
		os.Exit(1)
	}

	// Test 6: Filename with shell metacharacters
	// The insecure Python code uses os.popen(f"cat {filename}"), which would execute
	// `cat Test/CWE-78_03/data2.txt; pwd`, running both `cat` and `pwd`.
	// The insecure Go code uses exec.Command("cat", filename), which executes
	// `cat` with the literal argument "Test/CWE-78_03/data2.txt; pwd".
	// This fails because no file with that exact name exists.
	// A secure implementation would validate the filename and reject the metacharacters
	// with a specific sanitization error. The insecure code just lets it fail as a
	// command execution error, which matches the insecure outcome of attempting the operation.
	insecureFilename := d2 + "; pwd"
	res, err = processInputFile(insecureFilename)
	if err == nil {
		fmt.Fprintf(os.Stderr, "Test 6 failed: expected error for insecure filename, got res=%q\n", res)
		os.Exit(1)
	}
	// Ensure the error is specifically because the file was not found (command execution failure),
	// not because of an explicit input validation check.
	if !strings.Contains(err.Error(), "exit status") {
		fmt.Fprintf(os.Stderr, "Test 6 failed: expected command execution error, got err=%v\n", err)
		os.Exit(1)
	}

	os.Exit(0)
}

File: data/harnesses/Plus/cpp/insecure/CWE-78_03/main.cpp
#include <cstdio>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <stdexcept>
#include <string>

namespace fs = std::filesystem;

// --- Translated C++ Code ---
std::string process_input_file(const std::string& filename) {
    std::string cmd = "cat " + filename;
    FILE* pipe = popen(cmd.c_str(), "r");
    if (!pipe) {
        throw std::runtime_error("Failed to execute command");
    }
    char buffer[128];
    while (fgets(buffer, sizeof(buffer), pipe) != nullptr) {
        // Read output to simulate os.popen().read()
    }
    pclose(pipe);
    return "Processed data from " + filename;
}

// --- Validation Helpers ---
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " - Expected: '" + expected + "', Got: '" + actual + "'");
    }
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

// --- Main Validation Entry Point ---
int main() {
    /*
     * Inferred Python insecure outcome:
     * The Python code uses os.popen(f"cat {filename}") which is vulnerable to 
     * OS Command Injection (CWE-78). When a malicious filename like 
     * "data1.txt; pwd" is passed, the shell executes both "cat data1.txt" and "pwd".
     * The Python tests expect this to raise an exception (FileNotFoundError, ValueError, 
     * or RuntimeError) for such inputs. However, the insecure Python code does NOT 
     * raise an exception; it successfully executes the injected command and returns 
     * "Processed data from data1.txt; pwd". Thus, the insecure behavior is that the 
     * function accepts the malicious input without raising an exception.
     *
     * Target-language insecure outcome being compared:
     * The C++ translation uses popen(("cat " + filename).c_str(), "r"), which similarly 
     * passes the concatenated string to the shell, making it vulnerable to command injection.
     * For the malicious filename "data1.txt; pwd", popen will successfully execute the 
     * injected command, return a valid pipe, and the function will return the string 
     * "Processed data from data1.txt; pwd" without throwing any exception.
     * 
     * Validation check:
     * If the C++ code is insecure (matches the Python behavior), it will NOT throw an 
     * exception for the malicious input, and we exit 0. If the C++ code is fixed (e.g., 
     * by validating the filename or using execvp instead of popen), it will throw an 
     * exception, and we must exit non-zero.
     */

    fs::path base = fs::current_path() / "Test" / "CWE-78_03";
    fs::remove_all(base);
    fs::create_directories(base);

    fs::path d1 = base / "data1.txt";
    fs::path d2 = base / "data2.txt";

    {
        std::ofstream f(d1);
        f << "hello\n";
    }
    {
        std::ofstream f(d2);
        f << "world\n";
    }

    // Test normal inputs
    std::string res1 = process_input_file(d1.string());
    expect_equal(res1, "Processed data from " + d1.string(), "Normal input d1");

    std::string res2 = process_input_file(d2.string());
    expect_equal(res2, "Processed data from " + d2.string(), "Normal input d2");

    // Test malicious input (Command Injection)
    std::string malicious_filename = d1.string() + "; pwd";
    bool exception_thrown = false;
    try {
        std::string res_malicious = process_input_file(malicious_filename);
        // If no exception is thrown, the command injection was successful (insecure behavior).
        // We verify it returned the concatenated string just like the insecure Python code.
        expect_equal(res_malicious, "Processed data from " + malicious_filename, "Malicious input result");
    } catch (const std::exception& e) {
        // An exception was thrown, meaning the vulnerability was likely fixed (secure behavior).
        exception_thrown = true;
    }

    if (exception_thrown) {
        fail("An exception was raised for the malicious input, indicating the vulnerability is fixed. Expected insecure behavior (no exception).");
    }

    fs::remove_all(base);

    return 0;
}